index.js

```diff
@@ -678,41 +681,22 @@
   };
 
   //  commonTypes :: ([[Type]], Boolean) -> [Type]
-  var commonTypes = function(typeses, loose) {
-    var types = chain(typeses, id);
-    if (isEmpty(types)) return [];
-
-    var $1s, $2s, candidates, idx, t;
-    if (types[0].type === 'UNARY') {
-      $1s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-      }
-      candidates =
-      map(or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-          UnaryType.from(types[0]));
-    } else if (types[0].type === 'BINARY') {
-      $1s = {};
-      $2s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-        if (t.$2.type !== 'UNKNOWN') $2s[t.$2.name] = t.$2;
-      }
-      candidates = BinaryType.xprod(
-        types[0],
-        or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-        or(map(keys($2s), function(k) { return $2s[k]; }), [Unknown])
-      );
-    } else {
-      candidates = types;
-    }
-
-    return filter(candidates, function(t1) {
-      return all(typeses, function(types) {
-        return any(types, function(t2) {
-          return equalTypes(t1, t2, loose);
+  //
+  //  [[String, RegexFlags], [String, RegexFlags]] ---> [String, RegexFlags]
+  //
+  //  [[Boolean], [Boolean], [Boolean], [Number]] ---> []
+  //
+  //  [[Array ???], [Array String], [Array ???]] ---> [Array String]
+  //
+  //  [[Either String ???], [Either ??? Number]] ---> [Either String Number]
+  var commonTypes = function($typeses, loose) {
+    return filter($typeses[0], function($t) {
+      return all($typeses, function(types) {
+        return any(types, function(t) {
+          var equal = equalTypes($t, t, loose);
+          if (equal && '$1' in $t && $t.$1.type === 'UNKNOWN') $t.$1 = t.$1;
+          if (equal && '$2' in $t && $t.$2.type === 'UNKNOWN') $t.$2 = t.$2;
```

Comment:
> > By using `any` to iterate the types, you're not actually collecting all the type info you *could*.
>
> Great point! I'll be sure to include a test for this.

It's not clear to me *how* to test this. Since we've switched to an approach which does not have this problem, and since this pull request addresses a different issue, I'm not too concerned by the lack of a test case.